Port `circleci` join method to new join service

timothyb89 opened the pull request:
This ports the `circleci` join method to the new join service. The `circleci` package has been moved to `lib/join` with minimal changes, and a small compatibility shim has been added to reuse the same verification logic between the legacy and new entrypoints.

Testing included (on real CircleCI runs):
- Verified legacy clients can connect to new services via the compatibility layer
- Verified new clients can connect to legacy servers via the legacy joinclient
- Verified new clients can connect to new servers
- Verified invalid runs (contexts) can't pass verification with an incorrect organization/context ID

See also: [RFD 27e](https://github.com/gravitational/teleport.e/blob/master/rfd/0027e-auth-assigned-uuids.md)

timothyb89 committed:
Port `circleci` join method to new join service

This ports the `circleci` join method to the new join service. The
`circleci` package has been moved to `lib/join` with minimal changes,
and a small compatibility shim has been added to reuse the same
verification logic between the legacy and new entrypoints.

See also: [RFD 27e](https://github.com/gravitational/teleport.e/blob/master/rfd/0027e-auth-assigned-uuids.md)

timothyb89 committed:
Fix imports

timothyb89 committed:
Whitelist circleci in the new joinclient

nklaassen commented on lib/join/circleci/circleci.go:
```diff
@@ -0,0 +1,162 @@
+const IssuerURLTemplate = "https://oidc.circleci.com/org/%s"
+
+func issuerURL(template, organizationID string) string {
+	return fmt.Sprintf(template, organizationID)
```
kind of a nit but this is just string concatenation, no need to Sprintf here

However, we might want to be a bit more careful to avoid potential SSRF, even though organizationID comes from the token. Could build the URL using the net/url package instead, maybe something like

```go
url.URL{
	Scheme: "https",
	Host: "oidc.circleci.com",
}.JoinPath("org", organizationID).String()
```

timothyb89 committed:
SSRF-harden issuerURL and remove unnecessary clock

timothyb89 committed:
Merge remote-tracking branch 'origin/master' into timothyb89/new-join-service-circleci

timothyb89 committed:
Remove unnecessary circleci validator wrapper funciton

timothyb89 committed:
Move redefined errMockInvalid to spacelift for now

timothyb89 committed:
Fix issuerURL plumbing for tests

The change to remove template strings to harden against SSRF broke
tests that depended on this functionality. This reintroduces a
non-templated issuerURL, and leaves it to callers to call
`IssuerURL()`.

timothyb89 committed:
Don't export issuerURL

timothyb89 committed:
Merge remote-tracking branch 'origin/master' into timothyb89/new-join-service-circleci
